Overhaul audio section (#350)
gbdev/pandocs: src/Audio.md

A better description would be "chiptune": tunes produced by a chip.
This section describes how audio is generated on the Game Boy.

## Architecture

{{#include imgs/apu_overview.svg:2:}}

The Game Boy has four sound generation units, called **channels** 1 through 4, notated "CH1", "CH2", etc.
Unlike other sound chips, such as the SNES' S-DSP or the Atari 5200's POKEY, each sound channel is largely different from the others.

Each channel generates an electronic signal; these signals are then mixed into two new channels (for stereo: one for the left ear, one for the right ear), which are then individually amplified, and then output either to the headphone jack, or the speaker[^speaker_mono].

Channels 1 and 2, the "pulse channels", produce [pulse (rectangular) waves](https://en.wikipedia.org/wiki/Pulse_wave).
Channel 3, the "wave" channel, produces user-defined waves.
Channel 4 is the "[noise](https://en.wikipedia.org/wiki/Noise_in_music)" channel, producing a pseudo-random wave.

The VIN channel is a signal received directly from the cartridge, allowing external hardware to supply a fifth sound channel.

> No licensed games used this feature, and it was omitted from the Game Boy Advance.

Author: Got some links?

> No licensed games used this feature, and it was omitted from the Game Boy Advance.

Reviewer: Nanoloop: https://www.nanoloop.com/mono/index.html

PocketVoice:
http://n64.icequake.net/mirror/www.bung.com.hk/porducts/pocketvoice.htm
http://n64.icequake.net/mirror/www.bung.com.hk/support/download.htm

Nanoloop is still being sold, whereas PocketVoice hasn't been sold in about 20 years since Bung went under, and was probably one of their more obscure products. I think it's enough for now to just mention in passing that "no licensed games used the VIN pin, although two unlicensed cartridges, Bung PocketVoice, and Nanoloop Mono did."

Speaking of which... we shold probably also explain how...

- the VIN pin is sometimes used as a flash /WR pin.
- how it therefore often has a pullup resistor.
- how this means the behavior of the modulating VIN can produce different results depending on the cartridge it's running from.
- How one homebrew game (which I don't remember the name of) mistakenly used 'rNR50` as a 2x4 bit PCM register, because on the cart it was developed on the DC offset on VIN was just right for this to work ok, but won't always work as per the previous point.

> No licensed games used this feature, and it was omitted from the Game Boy Advance.

Author: Mm, honestly, I'd rather not touch VIN. This PR is enough of a brain space nightmare as-is. Open a separate issue, maybe?

::: tip POCKET MUSIC

> Despite rumors, *Pocket Music* does not use VIN. It refuses to run on the GBA for a different reason:

Author: I don't know, I believe @pinobatch added this line.

> Despite rumors, *Pocket Music* does not use VIN. It refuses to run on the GBA for a different reason:

Reviewer 2: I apologize for not having saved receipts when I first read this rumor.

Thus, the APU interfaces use durations instead of frequencies, which may be confusing as signal theory and music are more typically based on the latter.
Thus, durations will be expressed from their frequencies: for example, a "256 Hz tick" means "1/256th of a second".

The length of ticks is not affected by [CGB double speed](<#FF4D - KEY1 - CGB Mode Only - Prepare Speed Switch>), so the APU works just the same regardless of CPU speed.

::: warning

The Game Boy's APU is actually full of tricky details; this chapter will mostly describe the intended / common behavior, and often paper over bugs & quirks.
Readers wishing to learn more should read the [APU details](<#APU details>) chapter.

:::

### Envelope

Each channel's [**envelope**](https://en.wikipedia.org/wiki/Envelope_(music)) can be configured.
A signal's [envelope](https://en.wikipedia.org/wiki/Envelope_(waves)) is, roughly speaking, the variation of the signal's amplitude.
On the GB, "envelope" basically designates the channel's loudness.

All channels can have their envelope controlled to modulate their volume (CH3 less precisely than other channels).

> In addition, all channels but CH3 benefit from a functionality called **envelope sweep**, which automatically adjusts the envelope over time.

Author: Envelope is a modulation of amplitude, here a volume setting, so the terminology seems acceptable imo.

It's not possible to change this terminology, either, because it's used literally everywhere (including [hardware.inc](https://github.com/gbdev/hardware.inc/blob/05f5a9b6c7172abe1d7488080c1c050284c09226/hardware.inc#L230-L237)).

> In addition, all channels but CH3 benefit from a functionality called **envelope sweep**, which automatically adjusts the envelope over time.

Reviewer: The envelope is technically implemented as a sweep in the sense that it interpolates from one value to another. But hardware.inc, which copied the comment from nocash Pan Docs, seems to refer to the lower 3 bits of NRx2 as "Number of envelope sweep", and not the whole function of generating the envelope. In the overview, it was described like this with no mention of sweep (except referring to ch1's pitch sweep.)

>    Quadrangular wave patterns with sweep and envelope functions.
>    Quadrangular wave patterns with envelope functions.
>    Voluntary wave patterns from wave RAM.
>    White noise with an envelope function.

If anything, this seems like one of many quirks in Martin's English which should probably be improved instead of embraced.

Also note that nocash Pan Docs and hardware.inc don't refer to ch3 (`rNR32`) as having an envelope/being an envelope setting, but an output level.

> In addition, all channels but CH3 benefit from a functionality called **envelope sweep**, which automatically adjusts the envelope over time.

Author: Updated.

The parameters that can be controlled are how many times the envelope will be stepped, in which direction, and how long between steps.
Envelope sweep ticks at 64 Hz, and every 1&ndash;7 of those ticks, the envelope will be increased or decreased.

> Unfortunately, of [ADSR envelopes](https://en.wikipedia.org/wiki/ADSR_envelope), the Game Boy only provides for Attack and Sustain, via envelope sweep and sound length respectively.

Reviewer: This is not really correct. In envelope terms, the GB envelope generators would be either attack/sustain or just decay, depending on the direction. And while the sound length register could be used to emulate (effectively) sustain and instant release, it could also be used to stop the channel during the decay phase. I don't think it's necessarily a good idea to try to fit the GB envelope control into exact ADSR terminology, but rather try to explain what each function does on its own.

Author: Okay, if you say so.

### Sound length

All channels can be set to automatically shut themselves down after a certain amount of time.
This can be used to time the "sustain" part of ADSR, though the sound will cut fairly abruptly once the sound length "expires", instead of decaying.

A channel's sound length is initially set between 0 and 63, and, if enabled, ticks up[^len_cnt_dir] at 256 Hz (tied to [DIV-APU](<#DIV-APU>)).
When the counter reaches 64, the channel is turned off.

[^len_cnt_dir]:
Actually, the sound length is inverted when written, and *that* ticks down until it reaches 0.
But the effect is as if the counter ticked up.

> ### Wave length

Author: Ok with contracting.

What's the difference between the "conversion formula" and the "underlying function"?

Calling it "frequency" is outright wrong, since it's not what's being controlled, and has proven to cause confusion. This is why I'm changing it.

> ### Wave length

Reviewer: By conversion formula, I mean literally the equation "Frequency = 131072/(2048-x) Hz" or some variant of it. 

By underlying function, I mean (a description of) how and why the oscillator functions the way it does.

And I'm not suggesting that the value itself should be referred to as a frequency, since it's obviously not. What I'm saying is that the section should possibly be called frequency, and things should be explained in terms of controlling the frequency. This makes educational sense because it's describing the hardware starting form familiar territory. (Assuming you have knowledge about music/sound theory.)

As a first approximation, this could be described as a special pitch value used by the hardware which can be calculated from the frequency.

Then a later paragraph can describe the exact meaning of the value, and how it's used internally by the hardware.

> ### Wave length

Author: Resolved locally by using "Frequency" as the section name. The conversion formula is not provided in the overview, simply because it depends on the channel.

# Audio Overview

Game Boy audio is sometimes called "8-bit", but this doesn't mean much: the term is used to designate audio similar to that produced by 8-bit consoles.

> A better description would be "chiptune": tunes produced by a chip.

Reviewer: No! No! No! Please avoid the use of the word chiptune. It's one of the most controversial terms in the chiptune scene itself, when it comes to an exact definition. It can means anything from an Amiga module with short loopes, so called "chipped", samples (the origin of the term) to a purist view that it can only refer to music generated by hardware, to a more liberal view that it can mean music in a cetain style, allowing a song to be made with a Game Boy but not be "chiptune". In short it can mean anything and everything depending on who you ask, so in the end it means nothing.

Author: Do you have a better wording for that sentence, or should it be removed entirely?

Reviewer: Not sure. I guess both the first two sentences could be omitted, but then again, I'm not sure how much context you need to explain, in particular for younger readers. (God, I feel old...) Maybe a fair assumption is that anyone reading the document has heard at least a few example of music/sound produced by the Game Boy in games?

> Game Boy audio is sometimes called "8-bit", but this doesn't mean much: the term is used to designate audio similar to that produced by 8-bit consoles.

As for this sentence. It's true that 8-bit can be used to designate audio similar to that produced by 8-bit consoles (similar to how chiptune can mean anything and everything) but that sentence almost seems to imply that "8-bit" is used only to describe _similar_ sounds, not the actual thing itself.

Maybe something like?

> Game Boy audio is sometimes called "8-bit". This does not refer to the bit depth of the sound generated, but rather that it has sound capabilities typical of 8-bit consoles. Like much of its contemporary hardware, the Game Boy produces sound generated by simple digital circuits.

Reviewer 3: I like @nitro2k01 last suggestion.

This section describes how audio is generated on the Game Boy.

## Architecture

{{#include imgs/apu_overview.svg:2:}}

The Game Boy has four sound generation units, called **channels** 1 through 4, notated "CH1", "CH2", etc.

> Unlike some other sound chips, such as the SNES' S-DSP or the Atari 5200's POKEY, each sound channel is specialized in a way largely different from the other channels.

Reviewer: The SNES' S-DSP is one generation ahead. A better comparison might be the SID.

Each channel generates an electronic signal; these signals are then mixed into two new channels (for stereo: one for the left ear, one for the right ear), which are then individually amplified, and then output either to the headphone jack, or the speaker[^speaker_mono].

> Channels 1 and 2, the "pulse channels", produce [pulse (rectangular) waves](https://en.wikipedia.org/wiki/Pulse_wave).

Reviewer: Not sure about calling it rectangular wave. This seems like a more academic terminology, and is in my experience not commonly used in synthesis for music. (A legacy from Martin's "quadrangular" waves?) I would rather say something like "produce [pulse width modulated waves](https://en.wikipedia.org/wiki/Pulse-width_modulation) with 4 fixed pulse width settings."

Author: So this?
```suggestion
Channels 1 and 2, the "pulse channels", produce [pulse width modulated waves](https://en.wikipedia.org/wiki/Pulse-width_modulation) with 4 fixed pulse width settings.
```

Reviewer: Yes.

Channel 3, the "wave" channel, produces arbitrary user-supplied waves.
Channel 4 is the "[noise](https://en.wikipedia.org/wiki/Noise_in_music)" channel, producing a pseudo-random wave.

> The VIN channel is a signal received directly from the cartridge, allowing external hardware to supply a fifth sound channel.

Reviewer: I would specify "is an analog signal..." here.

Author: ```suggestion
The VIN channel is an analog signal received directly from the cartridge, allowing external hardware to supply a fifth sound channel.
```

No licensed games used this feature, and it was omitted from the Game Boy Advance.

::: tip POCKET MUSIC

Despite some online claims, *Pocket Music* does not use VIN.
It refuses to run on the GBA for a different reason: the developer couldn't figure out how to silence buzzing associated with sample playback on the wave channel.

:::

[^speaker_mono]:
The speaker merges back the two channels, losing the stereo aspect entirely.

## Common concepts

### APU

The Game Boy's sound chip is called the <abbr title="Audio Processing Unit">APU</abbr>.

The APU runs off the same master clock as the rest of the Game Boy, which is to say, it is fully synced with the CPU and [PPU](<#Rendering Overview>).
This also means that the APU runs about 2.4% faster on the SGB1, increasing frequencies by as much and thus sounding slightly higher-pitched.
The SGB2 rectifies this issue.

> The APU processes a lot of events at a fixed rate dependent on the event's type.

Author: Which rate it is depends on the event's type, but it's fixed for each type.
```suggestion
The APU processes a lot of events at fixed rates dependent on the event's type.
```
```suggestion
The APU processes a lot of events, at a different fixed rate depending on the event's type.
```

> The APU processes a lot of events at a fixed rate dependent on the event's type.

Reviewer: In my opinion, this whole sentence is at the wrong level of abstraction for this section of the document and should probably be removed entirely.

The length of APU ticks is not affected by [CGB double speed](<#FF4D — KEY1 (CGB Mode only): Prepare speed switch>), so the APU works just the same regardless of CPU speed.

::: warning

The Game Boy's APU is actually full of tricky details; this chapter will mostly describe the intended / common behavior, and often paper over bugs & quirks.
Readers wishing to learn more should read the [APU details](<#Audio Details>) chapter.

:::

### Triggering

**Triggering** a channel causes it to turn on if it wasn't[^trig_dac_off], and to start playing its wave from the beginning[^pulse_restart].
Most changes to a channel's parameters take effect immediately, but some require re-triggering the channel.

### Volume & envelope

The volume can be controlled in two ways: there is a "master volume" control[^vol_knob] (which has separate settings for the left and right outputs), and each channel's volume can be individually set as well (CH3's less precisely than the others).

Additionally, an [**envelope**](https://en.wikipedia.org/wiki/Envelope_(music)) can be configured for all channels except CH3, which allows automatically adjusting the volume over time.

> The parameters that can be controlled are how many times the envelope will be stepped, in which direction, and how long between steps.

Reviewer: Too heavy on technical primitives. I'd describe this as something like:
> The envelope register allows setting the initial volume, the envelope direction and envelope speed.

Author: I'd avoid the word "register" here too, then:
```suggestion
The parameters that can be controlled are the initial volume, the envelope's direction, and its speed.
```

Reviewer: Register is fine, since that's the way the programmer interfaces with the hardware, and something you need to learn early to do anything with the hardware. What I really mean by technical primitives, is _hidden_ technical primitives. Ie, new extra concepts you have to understand, just to read a basic sentence. With that said, there's nothing wrong with the suggested change, per se.

Author: I'm generally aiming for `Audio.md` to be readable by non-programmers, `Audio_Registers.md` by programmers, and `Audio_details.md` by people who really need to dive into some of the details (programmers trying to understand/debug weird behaviour, emu devs, etc.)

> All envelopes are ticked at 64 Hz, and every 1–7 of those ticks, the volume will be increased or decreased.

Reviewer: This is fine as is, after the above change, except maybe it should start with "Internally, all envelopes..."

### Length timer

All channels can be individually set to automatically shut themselves down after a certain amount of time.

If the functionality is enabled, a channel's **length timer** ticks up[^len_cnt_dir] at 256 Hz (tied to [DIV-APU](<#DIV-APU>)) from the value it's initially set at.
When the length timer reaches 64, the channel is turned off.

### Frequency

Music notes and audio waves are typically manipulated in terms of **frequency**[^pitch], i.e. how often the signal repeats per second.
However, as explained above, the Game Boy APU primarily works with durations; thus, **wavelengths** will be used instead of frequency.[^len_raw]

::: warning

The term "wavelength" throughout this document does *not* designate the inverse of the frequency, but instead a quantity akin to it.
See the description of each NRx3 register for more information.

:::

---

[^trig_dac_off]:

> If [the channel's DAC](#DACs) is off, the channel will not turn on.

Reviewer: This footnote makes little sense in the context it's linked from. "The channel will not turn on" in which situation? I think what this means to say is that the channel doesn't initially turn on just by writing other registers than bit 7 of the NRx4 but this doesn't come across.

Author: I meant that setting NRx4 bit 7 while the DAC is still off will not turn the channel on. The DAC must be turned on first, then the channel triggered.

Reviewer: Ok, I have to admit I was unaware of this particular fact: 
> Channel <var>x</var>'s DAC is enabled if and only if `[NRx2] & $F8 != 0`;"
I was under the impression that this was not a prerequisite for turning on the channel (in terms of for example making it's phase counter start counting). Is the research for this behavior documented anywhere?

Author: I don't know of any test ROM that checks this (maybe somewhere in blargg), but I just tested with [Numism's exerciser](https://github.com/pinobatch/numism/tree/main/gameboy/exercise#sound-exerciser), writing $80 to NR24 does nothing with NR22 = $00, but it does with e.g. $08.

[^pulse_restart]:
Except for pulse channels, whose index into their waves is only ever reset by turning the APU off.

> This is not noticeable when playing sound, and only matters for "advanced" APU usage.

Reviewer: Following the theme of not describing things solely in terms of primitives, "index into their waves" should probably be "phase" or "phase position". Also, I'm not sure this information should be hidden away in a footnote since it applies to half of the available channels. Also, not sure exactly what counts as advanced APU usage, but a fairly common trick is to play two slightly detuned notes on two channels, which can sound different every time due to this effect.

Author: This is my turn being confused—what does "detuned" mean?

I think "phase position" would best describe the intent here.

Reviewer: Detuned just means the two notes have a slightly different frequency. The effect in question happens when two such notes create a "phasing" sound with each other. A simple way to illustrate the concept is for example using my bgbtest.gb ROM (which comes with BGB.) Press one of the buttons quickly repeatedly. Every second note is detuned, and if you do it quick enough, they overlap and create this effect.

What you can't do in a very predictable way, is an effect similar to a PWM that goes from a low percent duty to a 50% duty, since the phase position might be "random" on every note retrigger. I can make a sound example later.

Reviewer 3: Agree here, I'd avoid saying "index into their waves"

Author: How is it now?

Reviewer 3: Ok to me, but I opened https://github.com/gbdev/pandocs/issues/445 to have the chance to continue this conversation

[^vol_knob]:
This is separate from the physical volume knob located on the side of the console.

[^len_cnt_dir]:
Internally, the length timer is inverted when written, and *that* ticks down until it reaches 0.
But the effect is as if the counter ticked up.

[^pitch]:
There is also **pitch**, which is merely a measure of how we perceive frequency.

> The higher the frequency, the higher the pitch; therefore, pitch will be omitted from the rest of the document.

Reviewer: Actually, rather than omitting pitch from the document, maybe using pitch as the preferred term instead of frequency is useful. Consider especially the noise channel, which for some settings doesn't have a clearly defined frequency, in terms of what the human ear can perceive. (For example 15-bit polynomial setting will produce a cycle that can be more than a second long.) Yet, it would be correct to say that we can vary the pitch of the sound.

Author: That doesn't mesh with the "registers" document, which expresses frequencies in terms of what is written to the registers. Pitch can be derived from that, but it's non-trivial.
